Flag PR for manual review on certain additions

bkjohnson opened the pull request:
## Description

This adds a script along with some CircleCI config that will run against a PR to see if certain configurable regexes match any code in the PR and flag them for manual review if a match is found.

The script is written in a functional style, so please let me know what your thoughts are on this.  I've done this because the `octokit` module I use for the Github API relies on Promises to begin with, and I also felt like functional programming would be easier to follow.

### Required Environment variables

The script relies on some environment variables.  If you run this locally, you will have to provide all of these.

#### Provided by Circle by default
-  `CIRCLE_PROJECT_REPONAME`
-  `CIRCLE_PROJECT_USERNAME`
-  `CIRCLE_PULL_REQUEST` (link to the PR -used to get PR number)

#### Configured in the Circle UI
- `BOT_NAME`
  - This isn't sensitive, but it made sense to pair it with the auth token
- `GITHUB_TOKEN`
  - An Oauth token used to make calls to the Github API

#### YAML config
- `CODE_PATTERN` 
    -  Regex pattern which will trigger a review comment if found
- `LINE_COMMENT`
  - Review comment for an individual line comment
- `OVERALL_REVIEW_COMMENT`
  - Review comment for the whole review

## Testing done

Local testing and testing on Circle's CI machines.


## Screenshots


## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bkjohnson committed:
See if circle sets PR number

bkjohnson committed:
Tweak syntax

bkjohnson committed:
Setup a workflow to run a pr-check on non-master

bkjohnson committed:
Indent config

bkjohnson committed:
Just try to run the job in the workflow

bkjohnson committed:
Try running a "build" before getting PR number

The docs [1] say that CIRCLE_PR_NUMBER and CIRCLE_PULL_REQUEST will be
available as environment variables, but based on a previous [2] that
doesn't seem to be the case.  The docs say that the environment
variables are:

> exported in each build

so maybe that means that they are only exported if the specific "build"
job gets run in a workflow.

[1] - https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
[2] - https://app.circleci.com/jobs/github/department-of-veterans-affairs/vets-website/19/parallel-runs/0/steps/0-102

bkjohnson committed:
Use list syntax for requires

bkjohnson committed:
Include colon

bkjohnson committed:
Try to run eslint-disable check in circle

Eventually we would want to limit this to happen only on branches with
PRs attached

bkjohnson committed:
Remove build from workflow

bkjohnson committed:
Add an eslint-disable

bkjohnson committed:
Remove the trailing colon

bkjohnson committed:
Checking for checks ;)

bkjohnson committed:
Try to leave a comment on each offending line

bkjohnson committed:
Install dependencies before running script

bkjohnson committed:
Try setting comments by offset instead

This will need some more explanation, but this is due to how the API
works

bkjohnson committed:
Set position to be 1 instead of 0

bkjohnson committed:
Use a map instead of a separate function

bkjohnson committed:
Delete old commented out code

This includes the pasted github actions code that was used as an example

bkjohnson committed:
Create function to encapsulate some things

This will help when we want to get line additions that aren't for
disabling eslint

bkjohnson committed:
Specify pattern from CI job

bkjohnson committed:
Update position even for deleted lines

This will help us use the right position when making the review comment
based off of position.  If the API changes so that we can use line
instead we will have to undo this.

bkjohnson committed:
Make review comments configurable

bkjohnson committed:
Add install-dependencies job as requirement

We can eventually use this for other workflows

bkjohnson committed:
Add install-dependencies to workflow

bkjohnson committed:
Go back to installing dependencies as a step

We can optimize this later.  This ended up breaking because the job
didn't have access to the code, so apparently `checkout` doesn't make
code available between jobs

bkjohnson committed:
Get the diff from the Github API

This may make more sense as the code gets tweaked.  Also, it feels more
"node-y" to use a REST API instead of using something cobbled together
using child_process

bkjohnson committed:
Refactor job so to be parameterized

This should allow us to use the same job to check for different things
in PRs

bkjohnson committed:
Turn jobs into mappable objects in config

bkjohnson committed:
Undo changes for parameterized jobs

This only seems available for version 2.1 - Not sure if we want to use
that yet

bkjohnson committed:
Add job to check for Sentry

bkjohnson committed:
Make Sentry call

bkjohnson committed:
Trigger Circle

bkjohnson committed:
Try to fix position reset

This is from Github's API documentation on the position for PR review
comments:

> The position value equals the number of lines down from the first "@@"
> hunk header in the file you want to add a comment. The line just
> below the "@@" line is position 1, the next line is position 2, and so
> on. The position in the diff continues to increase through lines of
> whitespace and additional hunks until the beginning of a new file.

Since the position continues to increase until a new file, I'll try
only resetting it when we reach a line that starts and ends with @@

bkjohnson committed:
Make sure position is updated on new chunk

If we reach the double @@ line in the same file, make sure position gets
updated

bkjohnson committed:
Translate add_lines from bash to JS

bkjohnson committed:
Remove grep child_process

bkjohnson committed:
Fix bug with position not resetting

bkjohnson committed:
Refactor for promise chaining

bkjohnson committed:
Remove a Sentry call

This is to test if the script still leaves a comment on the diff

bkjohnson committed:
Temporarily disable eslint check

bkjohnson committed:
Only include additions in output

This could have been achieved in the filtering step by prepending the
CODE_PATTERN search with a regex to match on the labels that have a "+"
at the end, but doing it earlier made more sense since we don't care
about the other lines anyway.

bkjohnson committed:
Add some comments

bkjohnson committed:
Don't add review comments for the same problems

Identifying the review comments that the "bot" left previously is a bit
tricky.  First, we fetch all the comments on a PR and ignore the ones
that are outdated (identified by having a null position).

This part is a bit hacky and I haven't thoroughly tested it yet - we
also ignore the comments that weren't made by a login equal to the
BOT_NAME environment variable.

This means that the environment variables for the token and the name
should be for the same account.

Now that we have only the comments that our review bot left, we can
determine if we have any new additions to make.  A new addition is an
addition which is on a new path/position combo and that has a new body.

bkjohnson committed:
Use correct body

The body in the stringified object is the comment text

bkjohnson committed:
Make sure Sentry & ESlint comments are different

If the comments are identical and someone had a call to Sentry and then
an inline comment at the end which disabled something in ESLint, then it
would create a race condition.  This kind of thing is unlikely to
happen, but we should try to guard against it.

bkjohnson committed:
Re-enable eslint check

bkjohnson committed:
Cast position to be int

The double quotes around the int was throwing off the stringified
comparison

bkjohnson committed:
Try to fix bug

bkjohnson committed:
Move function up top for readability

bkjohnson committed:
Exit if this run isn't for a PR

bkjohnson committed:
Remove unused install-dependencies job

bkjohnson committed:
Don't destructure arguments

This doesn't really benefit us here

bkjohnson commented on .circleci/config.yml:
```diff
@@ -34,4 +34,51 @@ jobs:
+    docker:
+      - image: circleci/node:10.18
+    environment:
+      CODE_PATTERN: (/* eslint-disable)|(// eslint-disable)
```
ESLint disabled here

bkjohnson commented on script/pr-check.js:
```diff
@@ -0,0 +1,151 @@
+/* eslint-disable camelcase */
```
ESLint disabled here

bkjohnson commented on script/pr-check.js:
```diff
@@ -0,0 +1,151 @@
+/* eslint-disable camelcase */
+/* eslint-disable no-console */
```
ESLint disabled here

bkjohnson commented on src/site/stages/build/process-cms-exports/schema-validation.js:
```diff
@@ -10,6 +10,9 @@ const { getContentModelType } = require('./helpers');
 const rawSchemasDir = path.join(__dirname, 'schemas', 'raw');
 const transformedSchemasDir = path.join(__dirname, 'schemas', 'transformed');
 
+// eslint-disable-next-line no-console
```
ESLint disabled here

bkjohnson reviewed (commented):
ESLint is disabled - manual review required.

bkjohnson committed:
Shift boundaries a bit

bkjohnson committed:
Remove old step

This was mostly for a sanity check

bkjohnson committed:
Don't exit early if additions is empty

I want to see if a review comment can be made if the comments array is
empty

bkjohnson reviewed (commented):
Sentry call found - Check for potential PII exposure

bkjohnson reviewed (commented):
ESLint is disabled - manual review required.

bkjohnson committed:
Add length check back in

bkjohnson committed:
Make beginning section a bit more readable

bkjohnson committed:
Refactor to create getPRbotComments function

bkjohnson committed:
Do some functional refactoring

bkjohnson committed:
Add a Sentry log back for testing

bkjohnson committed:
Do some more functional refactoring

bkjohnson commented on src/applications/hca/actions.js:
```diff
@@ -7,6 +8,8 @@ import {
   isEnrollmentStatusLoading,
 } from './selectors';
 
+Sentry.captureMessage('Something');
```
Sentry found

bkjohnson reviewed (commented):
Sentry call found - Check for potential PII exposure

bkjohnson committed:
Add some comments

bkjohnson committed:
Fix the "error" handling

bkjohnson committed:
Undo test changes

bkjohnson committed:
Modify circle config

bkjohnson committed:
Update the lockfile

bkjohnson committed:
Try to tag frontend-review-group

bkjohnson commented on src/applications/hca/actions.js:
```diff
@@ -30,6 +31,7 @@ export const SET_DISMISSED_HCA_NOTIFICATION = 'SET_DISMISSED_HCA_NOTIFICATION';
 export const SHOW_HCA_REAPPLY_CONTENT = 'SHOW_HCA_REAPPLY_CONTENT';
 
 export function showReapplyContent() {
+  Sentry.captureMessage('Something');
```
Sentry found

bkjohnson replied:
Sentry found

bkjohnson reviewed (commented):
Sentry call found - @frontend-review-group check for potential PII exposure

bkjohnson committed:
Try to create a review request

bkjohnson reviewed (commented):
Sentry call found - @frontend-review-group check for potential PII exposure

bkjohnson committed:
Undo some testing changes

bkjohnson committed:
Refactor labelAdditions a bit

I'm trying to make this more functional to be consistent with the rest
of the code

rianfowler commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+      // Only add to the output if this line of content is an addition (begins with a "+")
+      if (/^[+].*/.test(line)) output.push({ path, position, line });
+      // Increment position for each line of actual content in the diff
+      position++;
```
`reduce` might more idiomatic this this need to track / pass accumulators (output, position)

bkjohnson replied:
I toyed with making it a reducer - I may go back to that.  The tricky part was having it manage the state

rianfowler commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+ */
+function findPattern(arr) {
+  return Promise.resolve(
+    arr.filter(({ line }) => new RegExp(CODE_PATTERN).test(line)),
```
does filter return a promise? Is this async?

bkjohnson replied:
Filter returns a new array and isn't async by default.

rianfowler commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+  return Promise.resolve(
+    arr.filter(({ line }) => new RegExp(CODE_PATTERN).test(line)),
+  ).then(
+    patternList =>
```
is this a list of the patterns or pattern matches?

bkjohnson replied:
List of things that matched the pattern.  I'll rename to clarify.

rianfowler commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+    arr.filter(({ line }) => new RegExp(CODE_PATTERN).test(line)),
+  ).then(
+    patternList =>
+      isEmpty(patternList)
```
This is nice

rianfowler commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+      // Create sorted list of unique stringified objects for easy comparison
+      // Removing the duplicates shouldn't be necessary once everything is working properly
+      // because there shouldn't be any duplicates
+      sortBy(uniqBy(comments, JSON.stringify), JSON.stringify).map(
```
Is this the same thing is stringifying every item, adding them to a `Set`, then `Array.sort`ing them?

bkjohnson replied:
Good call.  I wanted to use a Set originally but it doesn't work for normal objects, but the stringified version should be fine.

rianfowler commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+}
+
+/**
+ * Receives a list of additions and returns a list containing those additions which (if any)
```
What is an addition?

bkjohnson replied:
An addition that appears in a diff, as identified by the `+` at the beginning of a line.

rianfowler commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+getPRdiff()
+  .then(labelAdditions)
+  .then(findPattern)
+  .then(filterAgainstPreviousComments)
```
Is this like a maybe add comment if it's not already added?

bkjohnson replied:
Exactly.  I think it could use a rename, but essentially we don't want to leave a comment on the same "offense" on the same line with the same comment if we've already done it before.

rianfowler reviewed (approved):
Looks good to me.

U-DON commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+
+/**
+ * Receives a list of additions and returns a list containing those additions which (if any)
+ * area considered "new" in this run.
```
```suggestion
 * are considered "new" in this run. 
```

U-DON commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+        JSON.stringify,
+      ),
+    )
+    .then(sortedUniqueFlat =>
```
Maybe something more descriptive like `uniqueBotComments`?

U-DON commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+      // Create sorted list of unique stringified objects for easy comparison
+      // Removing the duplicates shouldn't be necessary once everything is working properly
+      // because there shouldn't be any duplicates
+      sortBy(uniqBy(comments, JSON.stringify), JSON.stringify).map(
```
Would it be better to reverse the operations (`sortBy` and then `uniqBy` or [`sortedUniqBy`](https://lodash.com/docs/4.17.15#sortedUniqBy))?

U-DON commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+  const output = [];
+
+  diffLines.forEach(line => {
+    if (/--- (a\/)?.*/.test(line)) {
```
`git diff` has this format at the top of its output:

```
--- a/src/applications/hca/config/form.js
+++ b/src/applications/hca/config/form.js
```

So this and the next check are for these headings.

bkjohnson commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+      return;
+    }
+    const match = line.match(/\+\+\+ (b\/)?(.*$)/);
+    if (match) {
```
[Eugene's comment](https://github.com/department-of-veterans-affairs/vets-website/pull/11670#discussion_r378960921) is right.  This is checking to see if the line in a diff is the beginning of a file.

bkjohnson commented on script/pr-check.js:
```diff
@@ -0,0 +1,190 @@
+ * where the label consists of:
+ *   - filename
+ *   - offset from the top of the diff chunk
+ * And both of the label pieces are separated by a colon
```
No - I need to reword the comment.  Things have changed since I originally wrote it.

bkjohnson committed:
Fix typo in comment

Co-Authored-By: Eugene Doan <1067024+U-DON@users.noreply.github.com>

bkjohnson committed:
Rewrite comment for clarity & rename func

bkjohnson committed:
Improve readability of findPattern

bkjohnson committed:
Rename arg for clarity

bkjohnson committed:
Testing with bot token

bkjohnson commented on .circleci/config.yml:
```diff
@@ -35,3 +35,48 @@ jobs:
+    docker:
+      - image: circleci/node:10.18
+    environment:
+      CODE_PATTERN: (/* eslint-disable)|(// eslint-disable)
```
The regex pattern _does_ match itself here.  I'm not too concerned about it for now.  If it becomes an issue later on we can work on ignoring certain things or maybe changing the regex

bkjohnson committed:
Revert "Testing with bot token"

This reverts commit 27edb258ee179430f9bdc3cac4a3db7aea476447.

bkjohnson committed:
Switch to a reducer

bkjohnson committed:
Merge branch 'master' into bkjohnson/circleci-experiments2

bkjohnson committed:
Trigger Circle

bkjohnson commented:
I think the duplicate reviews were due to an improperly set environment variable in Circle CI.  I've reset it and rerun the jobs and it [finished with `No new comments to make`](https://app.circleci.com/jobs/github/department-of-veterans-affairs/vets-website/559), so I think we're good.  If we see this happen again I'll take another look.

bkjohnson committed:
Change wording to emphasize waiting for a review

bkjohnson committed:
Add initial draft for icon-check job

This is for accessibility reasons.  FontAwesome icons typically use the
<i> tag to add icons, and sometimes icons are used to convey meaning
that isn't captured in text.  This is what we want to be on the lookout
for.  In the cases where icons are used with a semantic purpose in mind,
we want to make sure that those semantics are conveyed to a screen
reader.

bkjohnson committed:
Reword comment string to avoid yaml error

bkjohnson committed:
Add another icon as a test

bkjohnson committed:
Be more specific with icon regex

bkjohnson committed:
Add another icon

bkjohnson committed:
Merge branch 'master' into bkjohnson/circleci-experiments2

bkjohnson committed:
Remove icon additions

These were for testing
